data/transactions/verify/txn_test.go

```diff
@@ -262,6 +264,143 @@ func TestDecodeNil(t *testing.T) {
 	}
 }
 
+type testDbgHook struct {
+	log []string
+}
+
+func (d *testDbgHook) BeforeTxn(ep *logic.EvalParams, groupIndex int) error {
+	d.log = append(d.log, "beforeTxn")
+	return nil
+}
+
+func (d *testDbgHook) BeforeLogicEval(cx *logic.EvalContext) error {
+	d.log = append(d.log, fmt.Sprintf("beforeLogicEval %s", cx.RunMode()))
+	return nil
+}
+
+func (d *testDbgHook) BeforeTealOp(cx *logic.EvalContext) error {
+	d.log = append(d.log, "beforeTealOp")
+	return nil
+}
+
+func (d *testDbgHook) BeforeInnerTxnGroup(ep *logic.EvalParams) error {
+	d.log = append(d.log, "beforeInnerTxnGroup")
+	return nil
+}
+
+func (d *testDbgHook) AfterInnerTxnGroup(ep *logic.EvalParams) error {
+	d.log = append(d.log, "afterInnerTxnGroup")
+	return nil
+}
+
+func (d *testDbgHook) AfterTealOp(cx *logic.EvalContext, evalError error) error {
+	d.log = append(d.log, "afterTealOp")
+	return nil
+}
+
+func (d *testDbgHook) AfterLogicEval(cx *logic.EvalContext, evalError error) error {
+	d.log = append(d.log, fmt.Sprintf("afterLogicEval %s", cx.RunMode()))
+	return nil
+}
+
+func (d *testDbgHook) AfterTxn(ep *logic.EvalParams, groupIndex int) error {
+	d.log = append(d.log, "afterTxn")
+	return nil
+}
+
+func TestTxnGroupWithDebugger(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	proto := config.Consensus[protocol.ConsensusCurrentVersion]
+
+	account := keypair()
+	accountAddr := basics.Address(account.SignatureVerifier)
+
+	// `#pragma version 6; int 1;`
+	program1 := []byte{0x06, 0x81, 0x01}
+	program1Addr := basics.Address(logic.HashProgram(program1))
+
+	// `#pragma version 6; byte "test"; pop; int 1;`
+	program2 := []byte{0x06, 0x80, 0x04, 0x74, 0x65, 0x73, 0x74, 0x48, 0x81, 0x01}
+	program2Addr := basics.Address(logic.HashProgram(program2))
+
+	makeHeader := func(sender basics.Address) transactions.Header {
```

Comment:
Consider txntest.Txn for easier construction of txns, including assembling the programs.